tidy3d/components/medium.py

```diff
@@ -3092,6 +3184,49 @@ def loss_upper_bound(self) -> float:
         ep = ep[~np.isnan(ep)]
         return max(ep.imag)
 
+    def compute_derivatives(self, derivative_info: DerivativeInfo) -> AutogradFieldMap:
+        """Compute adjoint derivatives for each of the ``fields`` given the multiplied E and D."""
+
+        # compute all derivatives beforehand
+        dJ_deps = self.derivative_eps_complex_volume(
+            E_der_map=derivative_info.E_der_map, bounds=derivative_info.bounds
+        )
+
+        dJ_deps = complex(dJ_deps)
+
+        # TODO: fix for multi-frequency
+        frequency = derivative_info.frequency
+        poles_complex = [(complex(a), complex(c)) for a, c in self.poles]
+        poles_complex = np.stack(poles_complex, axis=0)
+
+        # compute gradients of eps_model with respect to eps_inf and poles
+        grad_eps_model = ag.holomorphic_grad(self._eps_model, argnum=(0, 1))
+        with warnings.catch_warnings():
+            # ignore warnings about holmorphic grad being passed a non-complex input (poles)
+            warnings.simplefilter("ignore")
+            deps_deps_inf, deps_dpoles = grad_eps_model(
+                complex(self.eps_inf), poles_complex, complex(frequency)
+            )
+
+        # multiply with partial dJ/deps to give full gradients
+
+        dJ_deps_inf = dJ_deps * deps_deps_inf
+        dJ_dpoles = [(dJ_deps * a, dJ_deps * c) for (a, c) in deps_dpoles]
```

Comment:
Nitpick, but the explicit tuple isn't necessary, i.e. `for (a, c) in deps_dpoles` --> `for a, c in deps_dpoles`

Reply:
I can change it, but curious why it wouldn't be caught by ruff?

Reply:
Hmm yeah interesting. Parentheses are removed in tuple unpacking in for loops, but are left unchanged in  list comprehensions. The choice seems somewhat arbitrary...